### ISSUE TITLE
koji_tag_inheritance crash with "priority"

### ISSUE THREAD

**Issue by ktdreyer** [ktdreyer/koji-ansible#258]
When altering an existing inheritance relationship to set `maxdepth: 0`, the `koji_tag_inheritance` module gets a Fault back from the hub (hub version: 1.27.0).

Here's the change a playbook author made:

```diff
 - name: firefox-91-esr-stack build tag inheritance
   koji_tag_inheritance:
    child_tag: "rhel-{{R}}-firefox-91-esr-stack-build"
     parent_tag: "rhel-{{R}}-firefox-91-esr-stack-gate"
     priority: 0
+    maxdepth: 0
```

Here's the backtrace from Ansible:
```
File "modules/koji_tag_inheritance.py", line 342, in <module>
File "modules/koji_tag_inheritance.py", line 338, in main
File "modules/koji_tag_inheritance.py", line 322, in run_module
File "modules/koji_tag_inheritance.py", line 249, in add_tag_inheritance
File "lib64/python3.6/site-packages/koji/__init__.py", line 2310, in __call__
return self.__func(self.__name, args, opts)
File "lib64/python3.6/site-packages/koji/__init__.py", line 2740, in _callMethod
raise err
File "lib64/python3.6/site-packages/koji/__init__.py", line 2721, in _callMethod
return self._sendCall(handler, headers, request)
File "lib64/python3.6/site-packages/koji/__init__.py", line 2636, in _sendCall
return self._sendOneCall(handler, headers, request)
File "lib64/python3.6/site-packages/koji/__init__.py", line 2685, in _sendOneCall
ret = self._read_xmlrpc_response(r)
File "lib64/python3.6/site-packages/koji/__init__.py", line 2697, in _read_xmlrpc_response
result = u.close()
File "/usr/lib64/python3.6/xmlrpc/client.py", line 656, in close
raise Fault(**self._stack[0])
xmlrpc.client.Fault: <Fault 1: "<class 'KeyError'>: 'priority'">
```

**Comment by ktdreyer** [comment]
I've reproduced this in an integration test. Here's the minimal example:
```yaml
- koji_tag:
    name: example-parent
    state: present

- koji_tag:
    name: example-child
    state: present
    inheritance:
    - parent: example-parent
      priority: 0

# This crashes with <Fault 1: "<class 'KeyError'>: 'priority'">
- koji_tag_inheritance:
    parent_tag: example-parent
    child_tag: example-child
    priority: 0
    maxdepth: 0
```

I'm investigating further with a local hub to see what's going on server-side.

**Comment by ktdreyer** [comment]
Here's the server-side logs. On koji-hub-1.27.0-3.el8:

```
2022-01-05 18:29:20,372 [WARNING] m=setInheritanceData u=kdreyer p=727 r=192.168.121.1:57646 koji.xmlrpc: Traceback (most recent call last):
  File "/usr/share/koji-hub/kojixmlrpc.py", line 253, in _wrap_handler
    response = handler(environ)
  File "/usr/share/koji-hub/kojixmlrpc.py", line 277, in handle_rpc
    return self._dispatch(method, params)
  File "/usr/share/koji-hub/kojixmlrpc.py", line 314, in _dispatch
    ret = koji.util.call_with_argcheck(func, params, opts)
  File "/usr/lib/python3.6/site-packages/koji/util.py", line 271, in call_with_argcheck
    return func(*args, **kwargs)
  File "/usr/share/koji-hub/kojihub.py", line 11986, in setInheritanceData
    - group
  File "/usr/share/koji-hub/kojihub.py", line 709, in writeInheritanceData
    get_tag(parent_id, strict=True)
  File "/usr/share/koji-hub/kojihub.py", line 741, in _writeInheritanceData
    # check for duplicate priorities
KeyError: 'priority'
```

koji-hub-1.26.1-1.el8 has basically the same error, though line numbers shifted:

```
2022-01-05 18:33:00,208 [WARNING] m=setInheritanceData u=kdreyer p=3366 r=192.168.121.1:57654 koji.xmlrpc: Traceback (most recent call last):
  File "/usr/share/koji-hub/kojixmlrpc.py", line 253, in _wrap_handler
    response = handler(environ)
  File "/usr/share/koji-hub/kojixmlrpc.py", line 277, in handle_rpc
    return self._dispatch(method, params)
   File "/usr/share/koji-hub/kojixmlrpc.py", line 314, in _dispatch
    ret = koji.util.call_with_argcheck(func, params, opts)
  File "/usr/lib/python3.6/site-packages/koji/util.py", line 271, in call_with_argcheck
    return func(*args, **kwargs)
  File "/usr/share/koji-hub/kojihub.py", line 11825, in setInheritanceData
    return writeInheritanceData(tag, data, clear=clear)
  File "/usr/share/koji-hub/kojihub.py", line 692, in writeInheritanceData
    _writeInheritanceData(tag_id, changes, clear)
  File "/usr/share/koji-hub/kojihub.py", line 724, in _writeInheritanceData
    if orig[f] != link[f]:
KeyError: 'priority'
```

**Comment by ktdreyer** [comment]
`koji_tag_inheritance` calls `setInheritanceData(child_tag, new_rules)`.

The first rule has the special `"delete link": True` element. This tells the hub to remove the old rule and then add the new rule.

The problem is that Koji now prints an error if the `priority` integer is the same for the old rule and new rule.

koji-1.17.0 (March 2019) and koji-1.17.1 (Oct 2019) work. This is approximately the time at which we added support for editing attributes (like `maxdepth`) with `koji_tag_inheritance`.

Starting with koji-1.18.0 (Aug 2019), we get the following error:
```
GenericError: Changes should not contain duplicated parent_id(9)
```
This was introduced in https://pagure.io/koji/c/310786bac1dff6004caf99755c865e8e68274fb6


I investigated how the `koji` CLI does maxdepth manipulations on existing relationships. Example:
```
koji add-tag example-parent
koji add-tag example-child --parent example-parent
koji --debug --debug-xmlrpc edit-tag-inheritance --maxdepth=0 example-child example-parent
```
Interestingly that final `edit-tag-inheritance` command simply sends a single `setInheritanceData` call with the single new rule. It does not try to delete the existing rule.

**Comment by ktdreyer** [comment]
I've edited `koji_tag_inheritance` to send the single new rule to `setInheritanceData`, and this works.

Some other research questions I discovered:

Q: Does Koji's own codebase still use `"delete link"` any more?
A: Yes. It only does it when deleting existing links (CLI `remove-tag-inheritance` and web UI.), not editing them.

Q: Is there another RPC for removing inheritance links?
A: No, the only way is to call `setInheritanceData` with `"delete link"`.

Q: Did Koji ever use `"delete link"` to edit links?
A: I checked koji-1.17.1, and it did not. It only used that to delete links.
